cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,640 @@
+  // not push or pop from the stack while being in this state.
+  TT_STR,
+
+  // The state being active after encountering an escape symbol (e.g., '\'), while being in the
+  // TT_STR state.
+  TT_ESC,
+
+  // Total number of states
+  TT_NUM_STATES
+};
+
+/**
+ * @brief Definition of the symbol groups
+ */
+enum class DFASymbolGroupID : uint32_t {
+  OpenBrace,         ///< Opening brace SG: {
+  OpenBracket,       ///< Opening bracket SG: [
+  CloseBrace,        ///< Closing brace SG: }
+  CloseBracket,      ///< Closing bracket SG: ]
+  Quote,             ///< Quote character SG: "
+  Escape,            ///< Escape character SG: '\'
+  Other,             ///< SG implicitly matching all other characters
+  NUM_SYMBOL_GROUPS  ///< Total number of symbol groups
+};
+
+// The i-th string representing all the characters of a symbol group
+const std::vector<std::string> symbol_groups = {"{", "[", "}", "]", "\"", "\\"};
+
+// Transition table
+const std::vector<std::vector<StateT>> transition_table = {
+  /* IN_STATE         {       [       }       ]       "       \    OTHER */
+  /* TT_OOS    */ {TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS},
+  /* TT_STR    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_STR},
+  /* TT_ESC    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR}};
+
+// Translation table (i.e., for each transition, what are the symbols that we output)
+const std::vector<std::vector<std::vector<char>>> translation_table = {
+  /* IN_STATE        {      [      }      ]     "  \   OTHER */
+  /* TT_OOS    */ {{'{'}, {'['}, {'}'}, {']'}, {'x'}, {'x'}, {'x'}},
+  /* TT_STR    */ {{'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}},
+  /* TT_ESC    */ {{'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}}};
+
+// The DFA's starting state
+constexpr auto start_state = TT_OOS;
+}  // namespace to_stack_op
+
+//------------------------------------------------------------------------------
+// JSON TOKENIZER PUSHDOWN AUTOMATON
+//------------------------------------------------------------------------------
+namespace tokenizer_pda {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Symbol groups for the input alphabet for the pushdown automaton
+ */
+enum SGID : PdaSymbolGroupIdT {
+  /// Opening brace
+  OBC,
+  /// Opening bracket
+  OBT,
+  /// Closing brace
+  CBC,
+  /// Closing bracket
+  CBT,
+  /// Quote
+  QTE,
+  /// Escape
+  ESC,
+  /// Comma
+  CMA,
+  /// Colon
+  CLN,
+  /// Whitespace
+  WSP,
+  /// Other (any input symbol not assigned to one of the above symbol groups)
+  OTR,
+  /// Total number of symbol groups amongst which to differentiate
+  NUM_PDA_INPUT_SGS
+};
+
+/**
+ * @brief Symbols in the stack alphabet
+ */
+enum STACK_SGID : PdaStackSymbolGroupIdT {
+  /// Symbol representing the JSON-root (i.e., we're at nesting level '0')
+  STACK_ROOT = 0,
```

Comment:
See above.
```suggestion
enum class stack_symbol_group_id : PdaStackSymbolGroupIdT {
  /// Symbol representing the JSON-root (i.e., we're at nesting level '0')
  STACK_ROOT = 0,
```

Reply:
+1 for replacing `SGID` with full names.

Reply:
Yup, agree. Replaced with full name 👍